File: backend/pdd/classifier.py
"""
PDD Complexity Classifier — When to Suggest Part-Driven Design (F-013)
======================================================================

WHAT THIS FILE DOES
-------------------
Analyzes a user chat message and decides whether it is a simple single-part
request or complex enough to suggest Part-Driven Design (PDD) workflow.

MIRROR OF FRONTEND
------------------
Logic matches ``frontend/renderer/pdd/classifier.ts`` so backend RPC and UI
agree on the same classification rules.

CLASSIFICATIONS
---------------
- "simple" — model directly with Intent IR
- "suggest_pdd" — offer /vision and structured design docs first
"""

from __future__ import annotations

import re  # Regex for "100 x 50 mm" style dimension patterns
from typing import Any

# Words that hint at multi-part or system-level design
ASSEMBLY_KEYWORDS = [
    "assembly",
    "frame",
    "drone",
    "robot",
    "multi-part",
    "motor mount",
]
# e.g. "100 x 50 mm" or "10×20"
SIMPLE_DIMENSION_PATTERN = re.compile(r"\d+\s*(mm|cm|m|×|x)\s*\d+", re.IGNORECASE)


def classify_request_complexity(
    message: str,
    context: dict[str, Any] | None = None,
) -> dict[str, str]:
    """
    Return {classification, reason} for a chat message.

    Heuristics:
    - Assembly keywords → suggest_pdd
    - Long message without dimensions → suggest_pdd
    - Otherwise → simple
    """
    del context  # Reserved for future workspace-aware rules (part count, etc.)
    lower = message.lower()
    word_count = len(lower.split())

    has_assembly_cue = any(keyword in lower for keyword in ASSEMBLY_KEYWORDS)
    has_dimensions = bool(SIMPLE_DIMENSION_PATTERN.search(message))

    if has_assembly_cue or (word_count > 12 and not has_dimensions):
        return {
            "classification": "suggest_pdd",
            "reason": (
                "This request spans multiple parts or systems — "
                "Part-Driven Design can help structure it."
            ),
        }

    return {
        "classification": "simple",
        "reason": "Single-part request with enough structure to model directly.",
    }


def build_pdd_suggestion_payload(message: str) -> dict[str, Any]:
    """Build UI payload with action buttons for the PDD suggestion banner."""
    return {
        "kind": "pdd_suggestion",
        "message": message,
        "classification": "suggest_pdd",
        "actions": [
            {"label": "Start with /vision", "command": "/vision"},
            {"label": "Continue without PDD", "command": "Continue without PDD"},
        ],
    }

File: frontend/renderer/pdd/classifier.ts
/**
 * ============================================================================
 * FILE: classifier.ts — Client-side PDD complexity heuristics (F-013)
 * ============================================================================
 *
 * PDD = Part-Driven Design: a structured workflow for multi-part projects.
 * Before sending every message to the LLM, the UI can guess if the user is
 * asking for something simple (one part with dimensions) vs complex (assembly,
 * robot, frame). This mirrors backend logic so frontend tests stay consistent.
 * ============================================================================
 */

/**
 * ComplexityResult — outcome of classifyRequestComplexity().
 */
export interface ComplexityResult {
  // simple: model directly; suggest_pdd: offer /vision workflow.
  classification: "simple" | "suggest_pdd";
  // Human-readable explanation shown in the chat hint.
  reason: string;
}

/** Keywords that suggest multi-part / assembly work. */
const ASSEMBLY_KEYWORDS = ["assembly", "frame", "drone", "robot", "multi-part", "motor mount"];

/** Regex: looks for two numbers with units or × between them (e.g. "100×50 mm"). */
const SIMPLE_DIMENSION_PATTERN = /\d+\s*(mm|cm|m|×|x)\s*\d+/i;

/**
 * classifyRequestComplexity — cheap text heuristics on the user's chat message.
 */
export function classifyRequestComplexity(message: string): ComplexityResult {
  const lower = message.toLowerCase();
  const wordCount = lower.split(/\s+/).length;

  const hasAssemblyCue = ASSEMBLY_KEYWORDS.some((k) => lower.includes(k));
  const hasDimensions = SIMPLE_DIMENSION_PATTERN.test(message);

  // Long vague prompts without dimensions → suggest PDD; assembly words → suggest PDD.
  if (hasAssemblyCue || (wordCount > 12 && !hasDimensions)) {
    return {
      classification: "suggest_pdd",
      reason: "This request spans multiple parts or systems — Part-Driven Design can help structure it.",
    };
  }

  return {
    classification: "simple",
    reason: "Single-part request with enough structure to model directly.",
  };
}

/**
 * buildPddSuggestionActions — buttons the UI can show when PDD is suggested.
 */
export function buildPddSuggestionActions(): Array<{ label: string; command: string }> {
  return [
    { label: "Start with /vision", command: "/vision" },
    { label: "Continue without PDD", command: "Continue without PDD" },
  ];
}
